File: benchmarks/microbenchmarks/findemailwithattachment.cc
/**
 * Find all emails to a user containing attachment of a particular
 * type received in a particular year.
 */

#include <stdio.h>
#include <string.h>
#include <iostream>
#include <vector>
#include "pmgd.h"
#include "util.h"
#include "neighbor.h"
#include "time.h"

using namespace std;
using namespace PMGD;

/**
 * Program name.
 */
static const char prog_name[] = "findemailwithattachment";

/*
 * Tags and properties. 
 */
const char *person_str = "Person";
const char *email_str = "Email";
const char *to_str = "To";
const char *attachment_str = "Attachment";
const char *filename_str = "Filename";
const char *deliverytime_str = "DeliveryTime";
const char *messageid_str = "MessageId";

/**
 * Print information about how to invoke this program
 * @param o the put-to stream
 */
void print_usage(ostream &o)
{
    o << "Usage: " << prog_name << " [OPTION]... GRAPH RECIPIENT TYPE START END\n";
    o << "\n";
    o << "From GRAPH get a list of all emails sent between START and END to the given\n";
    o << "RECIPIENT with the email containing at least one attachment of the given\n";
    o << "TYPE.  TYPE is specified as a dotted suffix such as \".pdf\" or \".pptx\".\n";
    o << "START and END must be Windows-compliant date specification such as\n";
    o << "\"Wed Dec 31 06:25:48 PST 2003\".\n";
    o << "\n";
    o << "  -h  print this help and exit\n";
    o << "  -t  print elapsed time and usage\n";
    o << "  -v  be verbose and print the messages' unique identifiers\n";
}

/**
 * The application's main function
 * @param argc the number of command line arguments
 * @param argv the array of command line arguments
 * @return status of execution (0 okay, 1 bad arguments, 2 processing error)
 */
int main(int argc, char *argv[])
{
    bool timing = false;    ///< Enable timing measurement
    struct timeval t1, t2;
    struct rusage u1, u2;

    bool verbose = false;

    int num_emails;         ///< The number of emails

    // Parse the command line arguments
    int argi = 1;
    while (argi < argc && argv[argi][0] == '-') {
        if (strcmp(argv[argi], "-h") == 0) {
            print_usage(cout);
            return 0;
        }
        else if (strcmp(argv[argi], "-t") == 0)
            timing = true;
        else if (strcmp(argv[argi], "-v") == 0)
            verbose = true;
        else {
            fprintf(stderr, "%s: %s: Unrecognized option\n", prog_name, argv[argi]);
            return 1;
        }
        argi += 1;
    }
    if (argi + 1 > argc) {
        cerr << prog_name << ": No input graph\n";
        print_usage(cerr);
        return 1;
    }
    char *graph_name = argv[argi];
    argi++;
    if (argi + 1 > argc) {
        cerr << prog_name << ": No recipient's email address\n";
        print_usage(cerr);
        return 1;
    }
    char *recipient_email = argv[argi];
    argi++;
    if (argi + 1 > argc) {
        cerr << prog_name << ": No attachment type\n";
        print_usage(cerr);
        return 1;
    }
    char *attachment_type = argv[argi]; // e.g., ".pptx" or ".pdf"
    size_t attachment_type_len = strlen(attachment_type);
    argi++;
    if (argi + 1 > argc) {
        cerr << prog_name << ": No start of time range\n";
        print_usage(cerr);
        return 1;
    }
    struct tm tm;
    int hr, min;
    if (!string_to_tm(argv[argi], &tm, &hr, &min)) {
        cerr << prog_name << ": Malformed start of time range\n";
        print_usage(cerr);
        return 1;
    }
    Time start(&tm, hr, min);
    argi++;
    if (argi + 1 > argc) {
        cerr << prog_name << ": No end of time range\n";
        print_usage(cerr);
        return 1;
    }
    if (!string_to_tm(argv[argi], &tm, &hr, &min)) {
        cerr << prog_name << ": Malformed end of time range\n";
        print_usage(cerr);
        return 1;
    }
    Time end(&tm, hr, min);

    try {
        // Open the graph.
        Graph db(graph_name, Graph::ReadOnly);

        // Start a transaction.
        Transaction tx(db);

        // Convert C string to StringIDs.
        StringID person_sid(person_str);
        StringID email_sid(email_str);
        StringID to_sid(to_str);
        StringID attachment_sid(attachment_str);
        StringID filename_sid(filename_str);
        StringID deliverytime_sid(deliverytime_str);
        StringID messageid_sid(messageid_str);

        // Time and usage measurements.
        (void)getrusage(RUSAGE_SELF, &u1);
        (void)gettimeofday(&t1, NULL);

        // Initialize the number of recipients.
        num_emails = 0;
        vector<Node *> emails;

        try {
            // Get the node representing the recipient (we assume it is unique).
            NodeIterator ri = db.get_nodes(person_sid,
                                           PropertyPredicate(email_sid,
                                                             PropertyPredicate::Eq,
                                                             recipient_email));
            // If the recipient exists, look for messages.
            if (ri)
                for (NodeIterator mi = get_neighbors(*ri, Any, to_sid, false);
                     mi;
                     mi.next()) {
                    for (NodeIterator ai = get_neighbors(*mi, Outgoing, attachment_sid, false);
                         ai;
                         ai.next()) {
                        Property filenamep;
                        if (ai->check_property(filename_sid, filenamep)) {
                            string s = filenamep.string_value();
                            size_t len = s.length();
                            Property deliverytimep;
                            if (len >= attachment_type_len
                                && s.compare(len - attachment_type_len, attachment_type_len, attachment_type) == 0
                                && mi->check_property(deliverytime_sid, deliverytimep)) {
                                Time t = deliverytimep.time_value(); 
                                if (start <= t && t <= end) {
                                    emails.push_back(&*mi);
                                    num_emails++;
                                    break;
                                }
                            }
                        }
                    }
                }
        }
        catch (Exception e) {
            if (e.num != ReadOnly)
                throw;
        }

        // Time and usage measurements.
        (void)gettimeofday(&t2, NULL);
        (void)getrusage(RUSAGE_SELF, &u2);

        // Print a summary.
        cout << num_emails << "\n";

        // Print timing info if requested.
        if (timing) print_delta(cout, t1, t2, u1, u2);

        // Be more detailed if requested.
        if (verbose)
            for (auto n:emails)
                cout << n->get_property(messageid_sid).string_value() << "\n";
    }
    catch (Exception e) {
        print_exception(e, stderr);
        return 2;
    }

    return 0;
}

File: benchmarks/microbenchmarks/FindEmailWithAttachment.java
/**
 * Find all emails to a user containing attachment of a particular
 * type received in a particular year.
 */

import pmgd.*;
import java.util.Vector;
import java.lang.String;
import java.lang.management.*;
import java.text.SimpleDateFormat;
import java.text.ParseException;
import java.util.Date;

public class FindEmailWithAttachment {
    /**
     * Program name.
     */
    static String progName = "FindEmailWithAttachment";

    /*
     * Tags and properties. 
     */
    static String person_str = "Person";
    static String email_str = "Email";
    static String to_str = "To";
    static String attachment_str = "Attachment";
    static String filename_str = "Filename";
    static String deliverytime_str = "DeliveryTime";
    static String messageid_str = "MessageId";

    /**
     * The application's main function
     * @param argc the number of command line arguments
     * @param argv the array of command line arguments
     * @return status of execution (0 okay, 1 bad arguments, 2 processing error)
     */
    public static void main(String[] argv)
    {
        boolean timing = false;    ///< Enable timing measurement
        boolean verbose = false;
        int argc = argv.length;

        int num_emails;         ///< The number of emails

        // Parse the command line arguments
        int argi = 0;
        while (argi < argc && argv[argi].charAt(0) == '-') {
            if (argv[argi].compareTo("-h") == 0) {
                printUsage();
                return;
            }
            else if (argv[argi].compareTo("-t") == 0)
                timing = true;
            else if (argv[argi].compareTo("-v") == 0)
                verbose = true;
            else {
                System.out.println(progName + ": " + argv[argi] + ": Unrecognized option");
                return;
            }
            argi += 1;
        }
        if (argi + 1 > argc) {
            System.out.println(progName + ": No input graph");
            printUsage();
            return;
        }
        String graph_name = argv[argi];
        argi++;
        if (argi + 1 > argc) {
            System.out.println(progName + ": No recipient's email address");
            printUsage();
            return;
        }
        String recipient_email = argv[argi];
        argi++;
        if (argi + 1 > argc) {
            System.out.println(progName + ": No attachment type");
            printUsage();
            return;
        }
        String attachment_type = argv[argi]; // e.g., ".pptx" or ".pdf"
        argi++;
        if (argi + 1 > argc) {
            System.out.println(progName + ": No start of time range");
            printUsage();
            return;
        }

        Date start, end;
        SimpleDateFormat sdf = new SimpleDateFormat("EEE MMM dd HH:mm:ss zzz yyyy");
        try {
            start = sdf.parse(argv[argi]);
        } catch (ParseException pe) {
            System.out.println(progName + ": Unable to parse start date");
            return;
        }

        argi++;
        if (argi + 1 > argc) {
            System.out.println(progName + ": No end of time range");
            printUsage();
            return;
        }
        try {
            end = sdf.parse(argv[argi]);
        } catch (ParseException pe) {
            System.out.println(progName + ": Unable to parse end date");
            return;
        }
        ThreadMXBean bean = ManagementFactory.getThreadMXBean();
        long userTime, cpuTime;

        try {
            // Open the graph.
            Graph db = new Graph(graph_name, Graph.OpenOptions.ReadOnly);

            // Start a transaction.
            Transaction tx = new Transaction(db, false, true);

            // Convert C string to Strings.
            StringID person_sid = new StringID(person_str);
            StringID email_sid = new StringID(email_str);
            StringID to_sid = new StringID(to_str);
            StringID attachment_sid = new StringID(attachment_str);
            StringID filename_sid = new StringID(filename_str);
            StringID deliverytime_sid = new StringID(deliverytime_str);
            StringID messageid_sid = new StringID(messageid_str);

            // Initialize the number of recipients.
            num_emails = 0;
            Vector<Node> emails = new Vector<Node>();

            // Time measurement.
            long startTime = bean.getCurrentThreadUserTime();
            long startTotal = bean.getCurrentThreadCpuTime();

            try {
                // Get the node representing the recipient (we assume it is unique).
                NodeIterator ri = db.get_nodes(person_sid,
                        new PropertyPredicate(email_sid,
                            PropertyPredicate.Op.Eq,
                            new Property(recipient_email)), false);
                // If the recipient exists, look for messages.
                if (!ri.done()) {
                    for (NodeIterator mi = ri.get_current().get_neighbors(Node.Direction.Any, to_sid, false);
                            !mi.done();
                            mi.next()) {
                        for (NodeIterator ai = mi.get_current().get_neighbors(Node.Direction.Outgoing, attachment_sid, false);
                                !ai.done();
                                ai.next()) {
                            Property filenamep = ai.get_current().get_property(filename_sid);
                            if (filenamep != null) {
                                String s = filenamep.string_value();
                                Property deliverytimep;
                                if (s.endsWith(attachment_type) &&
                                    (deliverytimep = mi.get_current().get_property(deliverytime_sid)) != null) {
                                    Date t = deliverytimep.time_value().getDate();
                                    if (t.compareTo(start) >= 0 && t.compareTo(end) <= 0) {
                                        emails.add(mi.get_current());
                                        num_emails++;
                                        break;
                                    }
                                }
                            }
                        }
                    }
                }
            }
            catch (pmgd.Exception e) {
                e.print();
                return;
            }

            // Time measurement.
            long endTime = bean.getCurrentThreadUserTime();
            long endTotal = bean.getCurrentThreadCpuTime();
            userTime = (endTime - startTime)/1000;
            cpuTime = (endTotal - startTotal)/1000;

            // Print a summary.
            System.out.println(num_emails);

            // Print timing info if requested.
            if (timing) {
                long sysTime = cpuTime - userTime;
                float upercent = 100 * (userTime / (float)cpuTime);
                float spercent = 100 * (sysTime / (float)cpuTime);
                System.out.println("Elapsed time is " + cpuTime + " microseconds");
                System.out.println("User time is " + userTime + " microseconds (" + upercent + "%)");
                System.out.println("System time is " + sysTime + " microseconds (" + spercent + "%)");
            }

            // Be more detailed if requested.
            if (verbose)
                for (Node n:emails)
                    System.out.println(n.get_property(messageid_sid).string_value());
        }
        catch (pmgd.Exception e) {
            e.print();
            return;
        }

        return;
    }

    /**
     * Print information about how to invoke this program
     * @param o the put-to stream
     */
    static void printUsage()
    {
        System.out.println("Usage: " + progName + " [OPTION]... GRAPH RECIPIENT TYPE START END");
        System.out.println("From GRAPH get a list of all emails sent between START and END to the given");
        System.out.println("RECIPIENT with the email containing at least one attachment of the given");
        System.out.println("TYPE.  TYPE is specified as a dotted suffix such as \".pdf\" or \".pptx\".");
        System.out.println("START and END must be Windows-compliant date specification such as");
        System.out.println("\"Wed Dec 31 06:25:48 PST 2003\".");
        System.out.println("");
        System.out.println("  -h  print this help and exit");
        System.out.println("  -t  time the execution");
        System.out.println("  -v  be verbose and print the messages' unique identifiers");
    }

}
